### ISSUE TITLE
Session hang compiling ceres file

### ISSUE THREAD

**Issue by bbalaban** [ceres-solver/ceres-solver#1095]
I'm attempting a build of ceres 2.2 following instructions found here: 
http://ceres-solver.org/installation.html#using-ceres-with-cmake

Downloaded  ceres-solver-2.2.0.tar.gz
Running on an AWS t2.micro ec2, Ubuntu v22, 1cpu, 13gb memory, 30gb disk. Installed all the deps. Initial CMake ran fine. Ran make with VERBOSE=1 to capture the commands. Several files compile fine, then this command hangs:

cd /home/ubuntu/ceres-bin/internal/ceres && /usr/bin/c++ -DCERES_SUITESPARSE_VERSION=\"5.10.1\" -DGFLAGS_IS_A_DLL=0 -DGOOGLE_GLOG_DLL_DECL="" -DGOOGLE_GLOG_DLL_DECL_FOR_UNITTESTS="" -Dceres_EXPORTS -I/home/ubuntu/ceres-solver-2.2.0/include -I/home/ubuntu/ceres-solver-2.2.0/internal -I/home/ubuntu/ceres-bin/include -I/usr/include/suitesparse -I/usr/include/eigen3 -Wmissing-declarations -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-missing-field-initializers -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -MD -MT internal/ceres/CMakeFiles/ceres_internal.dir/generated/schur_eliminator_2_2_2.cc.o -MF CMakeFiles/ceres_internal.dir/generated/schur_eliminator_2_2_2.cc.o.d -o CMakeFiles/ceres_internal.dir/generated/schur_eliminator_2_2_2.cc.o -c /home/ubuntu/ceres-solver-2.2.0/internal/ceres/generated/schur_eliminator_2_2_2.cc

Initially cpu pegs at 100%, then drops to zero after 1-2 minutes. Console session is hung. Reboot repeat, same.

**Comment by sergiud** [comment]
The bottleneck is likely the RAM not the CPU. I suggest cross-compiling Ceres for your target system.

**Comment by bbalaban** [comment]
@sergiud Good call! Upgraded to a t2.large with 8gb ram, ran fine. Thanks!

**Comment by sergiud** [comment]
Just wondering: was "13gb memory" a typo and you actually meant to write "3gb memory". I'm asking because the new target would have even less RAM otherwise.

**Comment by bbalaban** [comment]
Yes, typo. Original config had 1gb RAM, new one 8GB. Funny the way it hung, though, with 0 cpu load. I would have expected huge swapping activity with a RAM bottleneck. But in any case it's good now. Might want to add a recommended RAM size in the doc.

**Comment by sergiud** [comment]
Thanks for the clarification.

Determining a recommended RAM size is not straightforward though as it highly depends on the (version of a) compiler, linker, operating system and environment, enabled set of Ceres features, dependencies, etc.

The general recommendation is therefore to cross-compile on a more powerful host system and afterwards deploy the artifacts to the target system. While such an approach requires a more elaborate setup to automate the process, the method allows for a shorter turnaround time.

**Comment by bbalaban** [comment]
Yes, that makes sense. Maybe a hint, though, of what symptoms one might see and what they mean. You were able to zone in on my root cause quick enough, but I spent a lot of time trying to figure out what (to me) looked like a mysterious issue.

**Comment by sergiud** [comment]
Good point.

@sandwichmaker We can at least mention in the docs that Ceres can fail to compile on systems with insufficient amount of RAM and users should cross-compile if increasing available memory is infeasible.
